[JSC] WASM IPInt SIMD: implement store, load, load and splat, load and extend

dhecht opened the pull request:
#### 12553bef37149a796dfaa74d3db1ccbd5cfefeb5
<pre>
[JSC] WASM IPInt SIMD: implement store, load, load and splat, load and extend
<a href="https://bugs.webkit.org/show_bug.cgi?id=298421">https://bugs.webkit.org/show_bug.cgi?id=298421</a>
<a href="https://rdar.apple.com/159898337">rdar://159898337</a>

Reviewed by Yusuke Suzuki.

For ARM, implement these instructions in WasmIPInt:

v128_load_mem
v128_load_8x8s_mem
v128_load_8x8u_mem
v128_load_16x4s_mem
v128_load_16x4u_mem
v128_load_32x2s_mem
v128_load_32x2u_mem
v128_load8_splat_mem
v128_load16_splat_mem
v128_load32_splat_mem
v128_load64_splat_mem
v128_store_mem

* JSTests/wasm/stress/simd-instructions-load-store.js: Added.
New test case to systematically verify each of these instructions in isolation

Canonical link: <a href="https://commits.webkit.org/299622@main">https://commits.webkit.org/299622@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66121b0cc88cb9623a2771901480d21d8851f932

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/119530 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/39222 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/29876 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/125804 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/71604 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9486a628-e298-4a07-9ae8-63df1c03cd45) 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/39920 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/47802 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/90796 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/60101 "Passed tests") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/6f37b39c-5afa-4d97-8114-8a129ce05415) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/122481 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/31859 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/107173 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/71288 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/c5fd45f5-328b-4719-bf38-11b8c5e3b9ee) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/30896 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/25277 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/69452 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/111657 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/101322 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/25470 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/128778 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/118048 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/46452 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Sequoia-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/99390 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/46817 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/103366 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/99218 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/25223 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/44649 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Sonoma-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/43016 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/46314 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/52020 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/146746 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/45778 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37727 "Found 1 new JSC binary failure: testapi, Found 18619 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/memop_bounds_check.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default, ChakraCore.yaml/ChakraCore/test/Function/StackArgsWithFormals.js.default ... (failure)") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/49129 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/47466 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/a90756301c7b66413377477f8b6f2ecd79069849. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50332

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/7e13b2ab3c7acda355500cad90d4de0ef5c1241d. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50332

webkit-early-warning-system commented:
Starting EWS tests for https://github.com/WebKit/WebKit/commit/66121b0cc88cb9623a2771901480d21d8851f932. Live statuses available at the PR page, https://github.com/WebKit/WebKit/pull/50332

dhecht committed:
[JSC] WASM IPInt SIMD: implement store, load, load and splat, load and extend
https://bugs.webkit.org/show_bug.cgi?id=298421
rdar://159898337

Reviewed by Yusuke Suzuki.

For ARM, implement these instructions in WasmIPInt:

v128_load_mem
v128_load_8x8s_mem
v128_load_8x8u_mem
v128_load_16x4s_mem
v128_load_16x4u_mem
v128_load_32x2s_mem
v128_load_32x2u_mem
v128_load8_splat_mem
v128_load16_splat_mem
v128_load32_splat_mem
v128_load64_splat_mem
v128_store_mem

* JSTests/wasm/stress/simd-instructions-load-store.js: Added.
New test case to systematically verify each of these instructions in isolation

Canonical link: https://commits.webkit.org/299622@main
